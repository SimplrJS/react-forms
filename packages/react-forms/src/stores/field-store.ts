import { BaseStore } from "../abstractions/base-store";
import { StoreHydrated } from "../actions/store-actions";
import { FieldValueChanged } from "../actions/field-actions";

export interface FieldStoreState {
    value: string;
}

export interface FieldStoreData {
    value: string;
}

export class FieldStore extends BaseStore<FieldStoreState, FieldStoreData> {
    constructor(protected readonly formId: string, protected readonly fieldId: string) {
        super();
    }

    protected getInitialState(): FieldStoreState {
        return {
            value: ""
        };
    }

    public updateValue(value: string): void {
        this.setState(new FieldValueChanged(value), state => ({
            ...state,
            value: value
        }));
    }

    public getValue(): string {
        return this.getState().value;
    }

    public hydrate(data: FieldStoreData): void {
        this.setState(new StoreHydrated(), state => data);
    }

    public dehydrate(): FieldStoreData {
        return this.getState();
    }
}
