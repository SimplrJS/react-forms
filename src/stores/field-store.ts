import { TinyEmitter, Callback } from "../helpers/emitter";

export class FieldStore extends TinyEmitter<Callback> {
    public initialValue: string;
    public defaultValue: string;
    public currentValue: string;

    /**
     *
     */
    constructor(defaultValue: string = "", initialValue?: string) {
        super();

        if (initialValue == null) {
            this.initialValue = defaultValue;
            this.currentValue = defaultValue;
        } else {
            this.initialValue = initialValue;
            this.currentValue = initialValue;
        }

        this.defaultValue = defaultValue;
    }

    public setCurrentValue(value: string): void {
        this.currentValue = value;
        this.emit();
    }
}
