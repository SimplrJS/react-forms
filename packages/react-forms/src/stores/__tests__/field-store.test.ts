import { FieldStore } from "../field-store";

const FAKE_FORMID = "fake-id";

it("Update value", () => {
    const store = new FieldStore(FAKE_FORMID);
    expect(store.getValue()).toMatchSnapshot();

    const nextValue = "Hello World!";
    store.updateValue(nextValue);

    expect(store.getValue()).toBe(nextValue);
});
