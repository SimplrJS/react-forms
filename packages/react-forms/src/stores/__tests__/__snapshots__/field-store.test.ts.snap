// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update value 1`] = `""`;
